src/index.ts

```diff
@@ -0,0 +1,188 @@
+let docCache: DocCache = {};
+
+interface FragmentSourceMap {
+  [framentName: string]: {
+    [source: string]: boolean;
+  };
+}
+
+// A map fragmentName -> [normalized source]
+let fragmentSourceMap: FragmentSourceMap = {};
+
+function cacheKeyFromLoc(loc?: Location): string {
+  if (!loc) return '';
+  return normalize(loc.source.body.substring(loc.start, loc.end));
+}
+
+// For testing.
+export function resetCaches(): void {
+  docCache = {};
+  fragmentSourceMap = {};
+}
+
+interface FragmentMap {
+  [key: string]: boolean
+}
+
+// Take a unstripped parsed document (query/mutation or even fragment), and
+// check all fragment definitions, checking for name->source uniqueness.
+// We also want to make sure only unique fragments exist in the document.
+let printFragmentWarnings = true;
+function processFragments(ast: DocumentNode): DocumentNode {
+  const astFragmentMap: FragmentMap = {};
+  const definitions: DefinitionNode[] = [];
+
+  for (let i = 0; i < ast.definitions.length; i++) {
+    const fragmentDefinition = ast.definitions[i];
+
+    if (fragmentDefinition.kind === 'FragmentDefinition') {
+      const fragmentName: string = fragmentDefinition.name.value;
+      const sourceKey: string = cacheKeyFromLoc(fragmentDefinition.loc);
+
+      // We know something about this fragment
+      if (fragmentSourceMap.hasOwnProperty(fragmentName) && !fragmentSourceMap[fragmentName][sourceKey]) {
+
+        // this is a problem because the app developer is trying to register another fragment with
+        // the same name as one previously registered. So, we tell them about it.
+        if (printFragmentWarnings) {
+          console.warn("Warning: fragment with name " + fragmentName + " already exists.\n"
+            + "graphql-tag enforces all fragment names across your application to be unique; read more about\n"
+            + "this in the docs: http://dev.apollodata.com/core/fragments.html#unique-names");
+        }
+
+        fragmentSourceMap[fragmentName][sourceKey] = true;
+
+      } else if (!fragmentSourceMap.hasOwnProperty(fragmentName)) {
+        fragmentSourceMap[fragmentName] = {};
+        fragmentSourceMap[fragmentName][sourceKey] = true;
+      }
+
+      if (!astFragmentMap[sourceKey]) {
+        astFragmentMap[sourceKey] = true;
+        definitions.push(fragmentDefinition);
+      }
+    } else {
+      definitions.push(fragmentDefinition);
+    }
+  }
+
+  // TODO: Ts won't allow us to alter this since it's read-only
+  (ast as any).definitions = definitions;
+  return ast;
+}
+
+export function disableFragmentWarnings(): void {
+  printFragmentWarnings = false;
+}
+
+function stripLoc(doc: Array<DocumentNode> | DocumentNode, removeLocAtThisLevel: number | boolean): DocumentNode  {
+  const docType = Object.prototype.toString.call(doc);
+
+  if (docType === '[object Array]' && doc instanceof Array) {
+    // This is wrong now because we return a DocumentNode... How do we approach this?
```

Comment:
```suggestion
    // TODO: This is wrong now because we return a DocumentNode... How do we approach this?
```

Reply:
yeah, this looks a little strange, but in this case the return value of this function is not accurate — when we get an array of DocumentNode, we return an array of DocumentNode